report/3.Theory/ObjectRecognition.tex

```diff
@@ -137,13 +137,97 @@ \subsection{Tensorflow}
+\textit{OpenCV was built to provide a common infrastructure for computer vision applications and to accelerate the use of machine perception in the commercial products\cite{opencvabout}.}
+\end{quote}
+According the above quote OpenCV should be easy to work with, and it should be easy and quick to implement different image manipulation tasks programmatically.
+Examples of tasks include\cite{opencvexamples}:
+\begin{itemize}[noitemsep]
+	\item Individual pixel manipulation, for example accessing all pixels of a certain color and changing them to another color, e.g. basic read/write.
+	\item Different forms of image processing, such as rotating the image, smoothing the image color space, image thresholding and more.
+	\item Feature and object detection, a more advanced image analysis that can find objects and features in image. 
+	Examples of usage here could be facial detection or detecting various objects in the image.
+\end{itemize}
+
+The below snippet is an example of facial recognition, implemented using OpenCV.
+\begin{lstlisting}[language=Python,label=lis:opencvExample,caption=OpenCV program to detect faces in an image.]
+	import numpy as np
+	import cv2
+	
+	face_cascade = cv2.CascadeClassifier('haarcascade_frontalface_default.xml')
+	eye_cascade = cv2.CascadeClassifier('haarcascade_eye.xml')
+	
+	img = cv2.imread('sachin.jpg')
+	gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
+	
+	faces = face_cascade.detectMultiScale(gray, 1.3, 5)
+	for (x, y, w, h) in faces:
+	  cv2.rectangle(img, (x, y), (x + w, y + h), (255, 0, 0), 2)
+	roi_gray = gray[y: y + h, x: x + w]
+	roi_color = img[y: y + h, x: x + w]
+	eyes = eye_cascade.detectMultiScale(roi_gray)
+	for (ex, ey, ew, eh) in eyes:
+	  cv2.rectangle(roi_color, (ex, ey), (ex + ew, ey + eh), (0, 255, 0), 2)
+	
+	cv2.imshow('img', img)
+	cv2.waitKey(0)
+	cv2.destroyAllWindows()
+\end{lstlisting}
+Snippet~\ref{lis:opencvExample} shows an example of simple facial recognition on an image, using Haar-Cascade.
+Haar-Cascade is a classifier, which has been trained to detect a specific object, in this case, faces.
+
+While OpenCV is not a machine learning library, it does include different machine learning elements, for example k-Nearest Neighbour (kNN) classification and K-Means clustering, based on images.
+kNN is a simple classification algorithm, that attempts to group different data into different classes, by looking at the data that most closely resembles each other.
+K-Means is another classification algorithm, that instead attempts to find a ''center'', a mean, and then simply assigning all the data that lies closest to this ''center'' to that class.
+
+The two algorithms share the trait, that they are often performed iteratively, meaning the same dataset are iterated over several times, leading to a more accurate classification, often by moving the classification of a single data-element.
+This shifts the ''weight'' of the class.
+
+Below is an example of ML implemented using OpenCV:
```

Comment:
Jeg kan bedre lide "snippet X is an example" derudover, "example of an ML algorithm". du ukan ikke implementere en læring. det er en abstraktion

Reply:
Første del er fordi jeg yderligere elaborater under snippeten. Anden del er en brain fart.